GitHubReferences.ps1

```diff
@@ -0,0 +1,442 @@
+        Gets all the references (everything in the namespace, not only heads and tags) if not provided
+
+    .PARAMETER AccessToken
+        If provided, this will be used as the AccessToken for authentication with the
+        REST Api. Otherwise, will attempt to use the configured value or will run unauthenticated.
+
+    .PARAMETER NoStatus
+        If this switch is specified, long-running commands will run on the main thread
+        with no commandline status update.  When not specified, those commands run in
+        the background, enabling the command prompt to provide status information.
+        If not supplied here, the DefaultNoStatus configuration property value will be used.
+
+    .OUTPUTS
+        [PSCustomObject]
+        Details of the git reference in the given repository
+
+    .EXAMPLE
+        Get-GitHubReference -OwnerName microsoft -RepositoryName PowerShellForGitHub -Reference heads/master
+#>
+    [CmdletBinding(
+        SupportsShouldProcess,
+        DefaultParametersetName='Elements')]
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSShouldProcess", "", Justification="Methods called within here make use of PSShouldProcess, and the switch is passed on to them inherently.")]
+    param(
+        [Parameter(ParameterSetName='Elements')]
+        [string] $OwnerName,
+
+        [Parameter(ParameterSetName='Elements')]
+        [string] $RepositoryName,
+
+        [Parameter(
+            Mandatory,
+            ParameterSetName='Uri')]
+        [string] $Uri,
+
+        [string] $Reference,
+
+        [string] $AccessToken,
+
+        [switch] $NoStatus
+    )
+
+    Write-InvocationLog -Invocation $MyInvocation
+
+    $elements = Resolve-RepositoryElements -BoundParameters $PSBoundParameters -DisableValidation
+    $OwnerName = $elements.ownerName
+    $RepositoryName = $elements.repositoryName
+
+    $telemetryProperties = @{
+        'OwnerName' = (Get-PiiSafeString -PlainText $OwnerName)
+        'RepositoryName' = (Get-PiiSafeString -PlainText $RepositoryName)
+        'ProvidedReference' = $PSBoundParameters.ContainsKey('Reference')
+    }
+
+    if ($OwnerName -xor $RepositoryName)
+    {
+        $message = 'You must specify both Owner Name and Repository Name.'
+        Write-Log -Message $message -Level Error
+        throw $message
+    }
+
+    if (-not [String]::IsNullOrEmpty($RepositoryName))
+    {
+        $uriFragment = "repos/$OwnerName/$RepositoryName/git/matching-refs"
+        $description =  "Getting all references for $RepositoryName"
+        if ($PSBoundParameters.ContainsKey('Reference'))
+        {
+            $uriFragment = $uriFragment + "/$Reference"
+            $description =  "Getting Reference $Reference for $RepositoryName"
+        }
+        else
+        {
+            # Add a slash at the end. Calling it with only matching-refs without the slash causes a 404
+            $uriFragment = $uriFragment + "//"

```

Comment:
If this is true, then the logic above is a bit delicate.  What if `$Reference` has a leading slash?  What happens then?  I think you need to do some more robust input validation.

Reply:
The additional slash at the end is required because Invoke-GHRestMethod removes trailing slashes.
https://github.com/microsoft/PowerShellForGitHub/blob/d279f1c7e1bd8261cdcf9dab6a8b78d3a0004457/GitHubCore.ps1#L126-L130
In this case, since there would not be any other variable appended, no input validation is required
And for the part where we do
`$uriFragment = $uriFragment + "/$Reference"`

Validation would not be required since the $reference variable would now be generated from the Tag/Branch passed in by the user

Please feel free to correct me if I misunderstood this comment